### PR TITLE
[system-probe] Route lookup traffic originating in non-root network namespace

### PR DESCRIPTION
**Pull request by hmahmood**
### What does this PR do?

- remove eBPF route lookup code since there is no good way to cache/lookup routes for all cases
- remove ebpfRouter
- modify netlinkRouter to use "iif" attribute to netlink route get request when necessary
- use nat'ed dest address for route lookup when applicable

### Motivation

What inspired you to submit this pull request?

### Additional Notes

Anything else we should know when reviewing?

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -227,3 +227,5 @@ replace (
 	k8s.io/metrics => k8s.io/metrics v0.18.2
 	k8s.io/sample-apiserver => k8s.io/sample-apiserver v0.0.0-20191016112829-06bb3c9d77c9
 )
+
+replace github.com/vishvananda/netlink => github.com/DataDog/netlink v1.0.1-0.20210312173533-6d628a7fc6f3
```
Comment (brycekahle) [inline go.mod:231]: Why is this fork necessary?

**Review by brycekahle — APPROVED** [review]
It seems OK to me. It would be great if you could add QA instructions. Trying to get your netlink change upstreamed would also be great.

File: pkg/network/route_cache.go [hunk 2]
```diff
@@ -119,22 +121,45 @@ func NewNetlinkRouter(procRoot string) (Router, error) {
...
+		log.Tracef("route for src=%s dst=%s: scope=%s gw=%+v if=%d", source, dest, r.Scope, r.Gw, r.LinkIndex)
 		return Route{
 			Gateway: util.AddressFromNetIP(r.Gw),
 			IfIndex: r.LinkIndex,
 		}, true
 	}
 
+	log.Tracef("could not get route for src=%s dest=%s err=%s routes=%+v", source, dest, err, routes)
+
 	return Route{}, false
```
Comment (p-lambert) [inline pkg/network/route_cache.go:164]: nit: I would do an early return for `if err != nil || len(routes) != 1` instead

File: pkg/network/tracer/tracer_test.go [hunk 3]
```diff
@@ -2200,7 +2210,148 @@ func TestGatewayLookupEnabled(t *testing.T) {
 	}, 2*time.Second, time.Second)
 
 	require.NotNil(t, conn.Via)
-	require.Equal(t, conn.Via.Subnet.Alias, fmt.Sprintf("subnet-%s", ifindex))
+	require.Equal(t, conn.Via.Subnet.Alias, fmt.Sprintf("subnet-%d", ifi.Index))
+}
+
+func TestGatewayLookupCrossNamespace(t *testing.T) {
```
Comment (p-lambert) [inline pkg/network/tracer/tracer_test.go:2216]: 👍

**Review by p-lambert — APPROVED** [review]

File: pkg/network/route_cache.go [hunk 4]
```diff
@@ -119,22 +121,45 @@ func NewNetlinkRouter(procRoot string) (Router, error) {
...
+	if n.rootNs != netns {
+		// if its a non-root ns, we're dealing with traffic from
+		// a container most likely, and so need to find out
+		// which interface is associated with the ns
+
+		// get input interface for src ip
+		routes, err := netlink.RouteGet(util.NetIPFromAddress(source))
+		if err != nil || len(routes) != 1 {
+			return Route{}, false
+		}
```
Comment (leeavital) [inline pkg/network/route_cache.go:134]: Is there going to be a case where `len(routes) > 1` ?

**Pull request merged**